File: 23/cups.py
from typing import List

inpt = "614752839"


def cups(labeling: str, moves: int = 100) -> str:
    cups = [int(c) for c in labeling]
    current = 0
    move = 0
    aside = []
    while (move < moves):
        move += 1
        currlabel = cups[current]
        # Take next three
        for _ in range(3):
            if (current + 1) < len(cups):
                aside.append(cups.pop(current + 1))
            else:
                aside.append(cups.pop(0))
                current -= 1
        # Find destination
        destlabel = cups[current] - 1
        while not destlabel in cups:
            destlabel -= 1
            if destlabel < min(cups):
                destlabel = max(cups)
                break
        destination = cups.index(destlabel)
        # Place in destination
        while len(aside):
            cups.insert(destination + 1, aside.pop())
        # Next current
        current = cups.index(currlabel) + 1
        if current == len(cups):
            current = 0

    # Find number one
    one = cups.index(1)
    return ''.join([str(x) for x in (cups[one + 1:] + cups[:one])])


print(f"The labels on the cups are {cups(inpt, 100)}")

File: 23/cups.cpp
#include <string.h>

#include <algorithm>
#include <iostream>
#include <list>
#include <queue>
#include <string>

#define INPUT "614752839" // "389125467"

void play(const std::string labeling, const int moves)
{
    std::list<int> cups;
    int maxCup = -1;
    for (char const &c : labeling)
    {
        const int cup = c - '0';
        cups.push_back(cup);
        if (cup > maxCup)
        {
            maxCup = cup;
        }
    }

    std::list<int>::iterator current = cups.begin();
    int move = 0;
    while (move++ < moves)
    {
        auto asideit = std::next(current, 1);
        std::queue<std::list<int>::iterator> aside;
        std::list<int> asidev;
        for (int i = 0; i < 3; i++)
        {
            if (asideit == cups.end())
            {
                asideit = cups.begin();
            }
            aside.push(asideit);
            asidev.push_back(*asideit);
            std::advance(asideit, 1);
        }

        int destLabel = *current - 1;
        int i = 0;
        while (i++ < 100)
        {
            if (destLabel == 0)
            {
                destLabel = maxCup;
            }
            if (std::find(asidev.begin(), asidev.end(), destLabel) != asidev.end())
            {
                destLabel--;
                continue;
            }
            auto destination = std::find(cups.begin(), cups.end(), destLabel);
            if (destination != cups.end())
            {
                std::advance(destination, 1);
                for (int i = 0; i < 3; i++)
                {
                    cups.splice(destination, cups, aside.front());
                    aside.pop();
                }
                break;
            }
            destLabel--;
        }

        std::advance(current, 1);
        if (current == cups.end())
        {
            current = cups.begin();
        }
    }

    auto one = std::find(cups.begin(), cups.end(), 1);
    auto it = std::next(one, 1);

    while (it != one)
    {
        if (it == cups.end())
        {
            it = cups.begin();
        }
        std::cout << *it;
        std::advance(it, 1);
    }
}

int main()
{
    std::cout << "The labels on the cups are ";
    play(INPUT, 100);
    std::cout << std::endl;

    return 0;
}
